Add [Nonlinear.ReverseAD] submodule (#1803)
jump-dev/MathOptInterface.jl: docs/src/submodules/Nonlinear/overview.md

[`eval_objective_gradient`](@ref) because [`Nonlinear.ExprGraphOnly`](@ref) does
not have the capability to differentiate a nonlinear expression.

If, instead, we pass [`Nonlinear.SparseReverseMode`](@ref), then we get access

> to `:Grad`, the gradient of the objective function, `:Jac`, the jacobian matrix

Reviewer: ```suggestion
to `:Grad`, the gradient of the objective function, `:Jac`, the Jacobian matrix
```

of the constraints, `:JacVec`, the ability to compute Jacobian-vector products,
and `:ExprGraph`.
```jldoctest nonlinear_developer
julia> evaluator = Nonlinear.Evaluator(
           model,
           Nonlinear.SparseReverseMode(),
           [x],
       )
Nonlinear.Evaluator with available features:
  * :Grad
  * :Jac
  * :JacVec
  * :ExprGraph
```

> However, before calling anything, we need to call [`initialize`](@ref):

Reviewer: ```suggestion
However, before using the evaluator, we need to call [`initialize`](@ref):
```

```jldoctest nonlinear_developer
julia> MOI.initialize(evaluator, [:Grad, :Jac, :JacVec, :ExprGraph])
```

Now we can call methods like [`eval_objective`](@ref):
```jldoctest nonlinear_developer
julia> x = [1.0]
1-element Vector{Float64}:
 1.0

julia> MOI.eval_objective(evaluator, x)
7.268073418273571
```
and [`eval_objective_gradient`](@ref):
```jldoctest nonlinear_developer

> julia> grad = [NaN]

Reviewer: Do we advise people to initialize vectors with `NaN`? I'd use zeros instead.

Author: I've been doing that in tests because it makes it clearer when you fail to hit some elements. But yes, in practice people should probably use zeros.

    Only call [`NLPBlockData`](@ref) once you have finished modifying the
    problem in `model`.

Putting everything together, you can create a nonlinear optimization problem in
MathOptInterface as follows:
```@example
import MathOptInterface
const MOI = MathOptInterface

> function build_model(model; backend)

Reviewer 2: repeat here what the type of the backend has to be here?
```suggestion
function build_model(model; backend::Nonlinear.AbstractAutomaticDifferentiation)
```

> function build_model(model; backend::Nonlinear.AbstractAutomaticDifferentiation)

Author: ```suggestion
function build_model(
    model::MOI.ModelLike;
    backend::MOI.Nonlinear.AbstractAutomaticDifferentiation,
)
```

[`Nonlinear.Model`](@ref) is a struct that stores the
[`Nonlinear.OperatorRegistry`](@ref), as well as a list of parameters and
subexpressions in the model.

## ReverseAD

`Nonlinear.ReverseAD` is a submodule for computing derivatives of a nonlinear
optimization problem using sparse reverse-mode automatic differentiation (AD).

This section does not attempt to explain how sparse reverse-mode AD works, but

> instead explains why MOI contains it's own implementation, and highlights

Reviewer: ```suggestion
instead explains why MOI contains its own implementation, and highlights
```

notable differences from similar packages.

!!! warning

>     You should not interact with `ReverseAD` directly. Instead, you should

Reviewer: Who does "you" refer to? It's not the case that "nobody" should interact with ReverseAD, otherwise why does the code exist? Be more specific, like "use [or don't use] X API to accomplish Y task".

[`Nonlinear.Model`](@ref) is a struct that stores the
[`Nonlinear.OperatorRegistry`](@ref), as well as a list of parameters and
subexpressions in the model.

## ReverseAD

`Nonlinear.ReverseAD` is a submodule for computing derivatives of a nonlinear
optimization problem using sparse reverse-mode automatic differentiation (AD).

This section does not attempt to explain how sparse reverse-mode AD works, but
instead explains why MOI contains it's own implementation, and highlights
notable differences from similar packages.

!!! warning
    You should not interact with `ReverseAD` directly. Instead, you should
    create a [`Nonlinear.Evaluator`](@ref) object with
    [`Nonlinear.SparseReverseMode`](@ref), and then query the MOI API methods.

> ### Why another AD package?

Reviewer: I reject the premise of the question because this code was one of the first implementations of reverse-mode AD in Julia. This discussion as worded doesn't really fit as package documentation. What would fit is a discussion of the design goals of `ReverseAD`. It's fair to mention the history of the code as well.

Author: How's this now? I only made minor changes to split the history and identify design goals. But can make larger changes if you want?

[`Nonlinear.Model`](@ref) is a struct that stores the
[`Nonlinear.OperatorRegistry`](@ref), as well as a list of parameters and
subexpressions in the model.

## ReverseAD

`Nonlinear.ReverseAD` is a submodule for computing derivatives of a nonlinear
optimization problem using sparse reverse-mode automatic differentiation (AD).

This section does not attempt to explain how sparse reverse-mode AD works, but
instead explains why MOI contains its own implementation, and highlights
notable differences from similar packages.

!!! warning
    Don't use the API in `ReverseAD` to compute derivatives. Instead, create a
    [`Nonlinear.Evaluator`](@ref) object with [`Nonlinear.SparseReverseMode`](@ref)
    as the backend, and then query the MOI API methods.

### Design goals

The JuliaDiff organization maintains a [list of packages](https://juliadiff.org)
for doing AD in Julia. At last count, there were at least ten packages–-not
including `ReverseAD`-–for reverse-mode AD in Julia. `ReverseAD` exists because
it has a different set of design goals.

 * **Goal: handle scale and sparsity**

>    The types of functions that MOI computes derivatives of have two key

Author: Done.
